src/report/ReportPoll.ts

```diff
@@ -0,0 +1,104 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Mjolnir, REPORT_POLL_EVENT_TYPE } from "../Mjolnir";
+import { ReportManager } from './ReportManager';
+import { LogLevel } from "matrix-bot-sdk";
+
+class InvalidStateError extends Error {}
+
+export class ReportPoll {
+    private from = 0;
+    private timeout: ReturnType<typeof setTimeout> | null = null;
+
+    /*
+     * A class to poll synapse's report endpoint, so we can act on new reports
+     *
+     * @param client The Matrix client underpinning the running Mjolnir
+     * @param manager The report manager in to which we feed new reports
+     * @param save An abstract function to persist where we got to in report reading
+     */
+    constructor(
+        private mjolnir: Mjolnir,
+        private manager: ReportManager,
+    ) { }
+
+    private schedulePoll() {
+        const self = this;
+        this.timeout = setTimeout(
+            function() { self.getAbuseReports() },
+            60_000 // a minute in milliseconds
+        );
+    }
+
+    private async getAbuseReports(): Promise<any> {
+        let response_: {
+            event_reports: { room_id: string, event_id: string, sender: string, reason: string }[],
+            next_token: number | undefined
+        } | undefined = undefined;
+        try {
+            response_ = await this.mjolnir.client.doRequest(
+                "GET",
+                "/_synapse/admin/v1/event_reports",
+                { from: this.from.toString() }
+            );
+        } catch (ex) {
+            await this.mjolnir.logMessage(LogLevel.ERROR, "getAbuseReports", `failed to poll events: ${ex}`);
+            this.schedulePoll();
+        }
+
+        const response = response_!;
+        for (let report of response.event_reports) {
+            let event: any; // `any` because `getEvent` and `handleServerAbuseReport` also use `any`
+            try {
+                event = await this.mjolnir.client.getEvent(report.room_id, report.event_id);
+            } catch (ex) {
+                continue;
+            }
+
+            await this.manager.handleServerAbuseReport({
+                roomId: report.room_id,
+                reporterId: report.sender,
+                event: event,
+                reason: report.reason,
+            });
+        }
+
+        if (response.next_token !== undefined) {
+            this.from = response.next_token;
+            await this.mjolnir.client.setAccountData(REPORT_POLL_EVENT_TYPE, { from: response.next_token });
```

Comment:
If this fails does polling for reports silently stop because we don't schedule another poll? We should probably make this scenario a bit louder. Do we also want it to fall over this way, and do we think that it can fall over often? Same goes for all awaits in `getAbuseReports`